### PR TITLE
Graphql & React Router 4

### PR DESCRIPTION
**Pull request by mmahalwy**
Fixes https://github.com/quran/quran.com-frontend/issues/675

Might fix https://github.com/quran/quran.com-frontend/issues/687
Might fix https://github.com/quran/quran.com-frontend/issues/830

Sorry for the large PR. Promise last one! But probably following PRs to polish this up before it goes to production.

Todos in following PRs:
- [ ] Audio segments
- [ ] Check and validate surah range
- [ ] Move to Jest for testing
- [ ] Break down files further, we should have very small fails to help with merge conflicts and org
- [ ] Write tests

Also included:
- Chunks are becoming smaller and smaller! This heavily cuts down files making the bundles really tiny
- Graphql for data
- React router v4, which is amazing for HMR

image

### CODE REVIEW

**Comment by mmahalwy** [comment]
Sorry for the large PR. Promise last one! But probably following PRs to polish this up before it goes to production.

**Comment by sabeurthabti** [comment]
iA tonight I will review - very exciting stuff.

**Comment by naveed-ahmad** [comment]
rebuild

File: src/containers/ChapterContainer/route.js [hunk 1]
```diff
@@ -0,0 +1,18 @@
+import loadable from 'loadable-components';
+
+export default [
+  {
+    path: '/:chapterId(\\d+)/:range/:translations',
+    component: loadable(
+      () => import(/* webpackChunkName: "ChapterContainer" */ './index') // import('./components/GlobalNav/Surah')
```
Comment (sabeurthabti) [inline src/containers/ChapterContainer/route.js]: **minor issue**

is there a better way to do these? seems like leaking web pack details is a smell.

Comment (mmahalwy) [reply src/containers/ChapterContainer/route.js]: @sabeurthabti sorry, do you mean leaking the chunk name?

Comment (mmahalwy) [reply src/containers/ChapterContainer/route.js]: It's not so bad to leak the chunk name. We do this commonly at Airbnb:

image

Helps us understand what chunk went wrong. That being said, we can change the webpack/prod.js file to output hashes

**Comment by mmahalwy** [comment]
rebuild

File: src/containers/ChapterContainer/route.js (hunk 1 above)
Comment (sabeurthabti) [reply src/containers/ChapterContainer/route.js]: I was mearly thinking from how easy we can make it for people who might not have been exposed to the concept of chunk names.

Comment (sabeurthabti) [reply src/containers/ChapterContainer/route.js]: agreed, it isn't so bad.

Comment (mmahalwy) [reply src/containers/ChapterContainer/route.js]: Apologies, I may not understand :(

**Comment by ahmedre** [comment]
Deployed to: http://staging.quran.com:32956

**Comment by mmahalwy** [comment]
Inshallah next I will add jest as the testing framework and add tests which will make me break down files further!

**Pull request merged**